qiskit/backends/ibmq/ibmqjob.py

```diff
@@ -0,0 +1,300 @@
+# =============================================================================
+"""IBMQJob module
+
+This module is used for creating asynchronous job objects for the
+IBM Q Experience.
+"""
+
+import time
+import logging
+import pprint
+
+from qiskit.backends import BaseJob
+from qiskit.backends.basejob import JobStatus
+from qiskit._qiskiterror import QISKitError
+from qiskit._result import Result
+from qiskit._resulterror import ResultError
+
+logger = logging.getLogger(__name__)
+
+
+class IBMQJob(BaseJob):
+    """IBM Q Job class"""
+
+    def __init__(self, q_job, api, submit_info, is_device):
+        """IBMQJob init function.
+
+        Args:
+            q_job (QuantumJob): job description
+            api (IBMQuantumExperience): IBM Q API
+            submit_info (dict): this is the dictionary that comes back when
+                submitting a job using the IBMQuantumExperience API.
+            is_device (bool): whether backend is a real device  # TODO: remove this after Qobj
+        """
+        super().__init__()
+        self._q_job = q_job
+        self._api = api
+        self._submit_info = submit_info
+        self._job_id = submit_info['id']
+        self._status = JobStatus.QUEUED
+        self._status_msg = None
+        self._cancelled = False
+        self._exception = None
+        self._is_device = is_device
+
+    def result(self, timeout=None, wait=5):
+        # pylint: disable=arguments-differ
+        this_result = self._wait_for_job(timeout=timeout, wait=wait)
+        if self._is_device and self.done:
+            _reorder_bits(this_result)  # TODO: remove this after Qobj
+        return this_result
+
+    def cancel(self):
+        """Attempt to cancel job. Currently this is only possible on
+        commercial systems.
+
+        Returns:
+            bool: True if job can be cancelled, else False.
+
+        Raises:
+            QISKitError: if server returned error
+        """
+        if self._is_commercial:
+            hub = self._api.config['hub']
+            group = self._api.config['group']
+            project = self._api.config['project']
+            response = self._api.cancel_job(self._job_id, hub, group, project)
+            if 'error' in response:
+                err_msg = response.get('error', '')
+                self._exception = QISKitError('Error cancelling job: %s' % err_msg)
+                raise QISKitError('Error canceelling job: %s' % err_msg)
+            else:
+                self._cancelled = True
+                return True
+        else:
+            # Since currently the API always reports RUNNING it isn't possible
+            # for the SDK to know when the job is actually running instead of just
+            # sitting in queue. Once that changes it may be useful to reimplement
+            # this class.
+            return False
+
+    @property
+    def status(self):
+        # check for submission error
+        if self._status == JobStatus.ERROR:
```

Comment:
Can you check if `self._status` can really have this value? It seems it is only written during `__init__` - if that is the case, it might be misleading to have a `self._status` field that does not really contain the current status of the job: can we either verify that the field always gets updated whenever the real status changes, or remove the field if it is not in use?

Reply:
I think conditional was not needed so it was removed. The way it's set up now, which seems to be the way futures is done, is that when an error happens `self._exception` is set to the `Exception` object and the exception is raised. I'm not sure if it is useful to do both of these things?